modules/F05_qemu_version_detection.sh

```diff
@@ -0,0 +1,103 @@
+#!/bin/bash
+
+# EMBA - EMBEDDED LINUX ANALYZER
+#
+# Copyright 2020-2021 Siemens Energy AG
+# Copyright 2020-2021 Siemens AG
+#
+# EMBA comes with ABSOLUTELY NO WARRANTY. This is free software, and you are
+# welcome to redistribute it under the terms of the GNU General Public License.
+# See LICENSE file for usage of this software.
+#
+# EMBA is licensed under GPLv3
+#
+# Author(s): Michael Messner, Pascal Eckmann
+
+# Description:  This module extracts version information from the results of S115
+
+F05_qemu_version_detection() {
+  module_log_init "${FUNCNAME[0]}"
+  module_title "Identified software components - via usermode emulation."
+
+  LOG_PATH_S115="$LOG_DIR"/s115_usermode_emulator.txt
+  if [[ -f "$LOG_PATH_S115" ]]; then
+    LOG_PATH_MODULE_S115="$LOG_DIR"/s115_usermode_emulator/
+
+    while read -r VERSION_LINE; do 
+      if echo "$VERSION_LINE" | grep -v -q "^[^#*/;]"; then
+        continue
+      fi
+
+      if [[ $THREADING -eq 1 ]]; then
+        version_detection_thread &
+        WAIT_PIDS_F05+=( "$!" )
+      else
+        version_detection_thread
+      fi
+    done < "$CONFIG_DIR"/bin_version_strings.cfg
+    echo
+    if [[ $THREADED -eq 1 ]]; then
+      wait_for_pid "${WAIT_PIDS_F05[@]}"
+    fi
+  fi
+
+  module_end_log "${FUNCNAME[0]}" "$QEMULATION"
+}
+
+version_detection_thread() {
+  BINARY="$(echo "$VERSION_LINE" | cut -d: -f1)"
+  STRICT="$(echo "$VERSION_LINE" | cut -d: -f2)"
+  LIC="$(echo "$VERSION_LINE" | cut -d: -f3)"
+
+  VERSION_IDENTIFIER="$(echo "$VERSION_LINE" | cut -d: -f4- | sed s/^\"// | sed s/\"$//)"
+
+  # if we have the key strict this version identifier only works for the defined binary and is not generic!
+  if [[ $STRICT != "strict" ]]; then
+    readarray -t VERSIONS_DETECTED < <(grep -a -o -H -E "$VERSION_IDENTIFIER" "$LOG_PATH_MODULE_S115"/qemu_*.txt | sort -u 2>/dev/null)
+  else
+    if [[ -f "$LOG_PATH_MODULE_S115"/qemu_"$BINARY".txt ]]; then
+      VERSION_STRICT=$(grep -a -o -E "$VERSION_IDENTIFIER" "$LOG_PATH_MODULE_S115"/qemu_"$BINARY".txt | sort -u | head -1 2>/dev/null)
+      BINARY_PATH=$(grep -a "Emulating binary:" "$LOG_PATH_MODULE_S115"/qemu_"$BINARY".txt | cut -d: -f2 | sed -e 's/^\ //' | sort -u | head -1 2>/dev/null)
+      if [[ -n "$VERSION_STRICT" ]]; then
+        if [[ "$BINARY" == "smbd" ]]; then
+          # we log it as the original binary and the samba binary name
+          VERSION_="$BINARY $VERSION_STRICT"
+          VERSIONS_DETECTED+=("$VERSION_")
+          BINARY="samba"
```

Comment:
At the moment there is only one exception, therefore I am fine with hard coding it. If the number of exceptions are increasing, we have to put these into a config file.